service_crategen/src/codegen/generator/rest_json.rs

```diff
@@ -364,11 +356,110 @@ fn generate_documentation(operation: &Operation) -> Option<String> {
+/// the fields in the result object.
+///
+/// Needs to determine whether or not other fields in the result object
+/// will be set later (e.g. from headers), so the compiler won't spit out
+/// warnings about unnecessary mutability
+fn generate_body_parser(operation: &Operation, service: &Service) -> String {
+
+    if operation.output.is_none() {
+        return "let result = ();".to_string();
+    }
+
+    let shape_name = &operation.output.as_ref().unwrap().shape;
+    let output_shape = &service.shapes()[shape_name];
+
+    let mutable_result = output_shape.members
+                .as_ref()
+                .unwrap()
+                .iter()
+                .any(|(_, member)| member.location.is_some());
+
+    match output_shape.payload {
+        None => json_body_parser(shape_name, mutable_result),
+        Some(ref payload_member_name) => {
+            let payload_member_shape = &output_shape.members.as_ref().unwrap()[payload_member_name].shape;
+            let payload_shape = &service.shapes()[payload_member_shape];
+            match payload_shape.shape_type {
+                payload_type if payload_type == ShapeType::Blob ||
+                                payload_type == ShapeType::String => {
+                    payload_body_parser(payload_type, shape_name, payload_member_name, mutable_result)
+                }
+                _ => json_body_parser(shape_name, mutable_result),
+            }
+        }
+    }
+}
+
+/// Take the raw http response body and assign it to the payload field
+/// on the result object
+fn payload_body_parser(payload_type: ShapeType, output_shape: &str, payload_member: &str, mutable_result: bool) -> String {
+
+    let response_body = match payload_type {
+        ShapeType::Blob => "Some(response.body)",
+        _ => "Some(String::from_utf8_lossy(&response.body).into_owned())",
+    };
+
+    format!("
+        let {mutable} result = {output_shape}::default();
+        result.{payload_member} = {response_body};
+        ",
+        output_shape = output_shape,
+        payload_member = payload_member.to_snake_case(),
+        response_body = response_body,
+        mutable = if mutable_result { "mut" } else { "" }
+    )
+}
+
+/// Parse the http response body as a JSON object with serde, and use that
+/// as the result object
+fn json_body_parser(output_shape: &str, mutable_result: bool) -> String {
+    // `serde-json` serializes field-less structs as "null", but AWS returns
+    // "{{}}" for a field-less response, so we must check for this result
+    // and convert it if necessary.
+    format!("
+            let mut body = response.body;
+
+            if body == b\"{{}}\" {{
+                body = b\"null\".to_vec();
+            }}
+
+            debug!(\"Response body: {{:?}}\", body);
+            debug!(\"Response status: {{}}\", response.status);
+            let {mutable} result = serde_json::from_str::<{output_shape}>(String::from_utf8_lossy(&body).as_ref()).unwrap();
```

Comment:
Should we be using `from_slice` on the response body bytes directly instead of lossily converting to a String first?